### PR TITLE
STN-628:  Caption/credit text styles limit

### PR DESCRIPTION
**Pull request by rkelbel**
# READY FOR REVIEW

## Summary
Limit the Caption/Credit Text Styles that print to prevent Site Owners from breaking the caption/credits view mode display styles with additional styling. Also adjusting for styles and elements that may be brought into the credits section from Drupal 7 that has unwanted styles or elements.

**Implementation Details:**
We limited the caption/credits field to accept Minimal HTML, which includes the following styles `<em> <strong> <cite> <blockquote cite> <code> <ul type> <ol start type> <li> <dl> <dt> <dd> <h2 id> <h3 id> <h4 id> <h5 id> <h6 id> <hr> <p class=""> <span class=""> <table class=""> <br> <a href hreflang data-entity-substitution data-entity-type data-entity-uuid title class=" hs-button hs-button--big hs-secondary-button hs-more-link hs-external-link hs-file-link hs-pdf-link hs-private-link">`

We would like to allow the following styles (for all other styles we do not want to print the styles, but print the content) for the view mode:
```
<em>
<strong>
<p>
<br>
<a>
```

**To accomplish this we need to:**
Target all text styles not in the above list and use the mixin `hb-text-reset`
Target all non text styles in the above list and set `display: none;`

My approach:
I wanted to utilize the `hb-text-reset` for all the unwanted elements and styles and then add in what we would like to allow. Those items include not just the elements we want to keep but allowing elements to keep their base style identity and mix in with the credits default look. The `<a>` tags to still be identified as links while still keeping the accessibility needed within the dark gradient overlay of the credits area. The strong tag, was reset but added a bold weight back to it to keep it's emphasized identity. Some elements were easy to remove such as `<hr>`, but most others needed to be reset and the content retained, which required some heavy-handedness in a couple of areas as well, so we did have to add a couple of`!important` attributes for those items.

Additionally, since we were adding padding originally to the p tag within the caption container, the overall look was reliant upon that. I reworked this now so the container is taking care of the padding and the child elements are able to display inline next to one another.  Additionally, since the child elements float: right, I added a `flex-end` with `display: flex` to the container so that the elements will display in the order they were added to the WYSIWYG.

Jira Card: [#628](https://sparkbox.atlassian.net/browse/STN-628)

## Need Review By (Date)
01/25

## Urgency
'Medium'

## Steps to Test
1. `npm run test`
2. To test this PR you will first need to turn on the caption/credits to display in the View Mode. You can find this [here](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/display), making sure the "Caption/Credit" field is enabled.  Additionally you may need to make sure it enabled in the [Form Display](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/form-display) if is not already.
3. You will also want to turn on FULL HTML for the credit/caption WYSIWYG even though this option will not be enabled for users later, it will allow you to enter these elements for testing using the style buttons as well as Source access. [You can add the FULL HTML option to credit/captions here.](http://sparkbox-sandbox.suhumsci.loc/admin/structure/media/manage/image/fields/media.image.field_media_image_caption)
4. Once those items are turned on you can start adding credits/captions to your media images you will be testing. The easiest method for this is to go to the [Media Grid page](http://sparkbox-sandbox.suhumsci.loc/admin/content/media-grid) and choose any images you are going to use in testing, switch your WYSIWYG entry to full and add in your elements on the credit/caption field. Doing this here will allow you to choose your images you've edited when adding postcards. I tested two or three images in postcards at a time with different elements.
5. Now, all you have left to do is to[ create Flexible Page](http://sparkbox-sandbox.suhumsci.loc/node/add/hs_basic_page) and add as many postcards to the page with your media images you've added caption/credits to for testing.
I've included some html elements that I used for testing below for easy copy and pasting into Source input, if you'd like to use it:

```
<h1>This is an H1</h1>
<h2>This is an H2</h2>
<h3>This is an H3</h3>
<h4 id="">This is an H4</h4>
<h5 id="">This is an H5</h5>
<h6 id="">This is an H6</h6>
<p>This is a standard caption p tag.</p>
<p><span>This is a span</span></p>
<p><a href="http://google.com">This is an anchor</a></p>
<p><strong>This is strong</strong></p>
<p><a class="hs-pdf-link" href="&lt;front&gt;">This is a pdf link</a></p>
<p><a class="hs-file-link" href="&lt;front&gt;">This is a file link</a></p>
<p><a class="hs-private-link" href="&lt;front&gt;">This is a private link</a></p>
<p><a class="hs-more-link" href="&lt;front&gt;">This is a more link</a></p>
<p><a class="hs-secondary-button" href="&lt;front&gt;">This is a secondary button</a></p>
<p><a class="hs-button--big" href="&lt;front&gt;">This is a big primary button</a></p>
<p><a class="hs-button" href="&lt;front&gt;">This is a primary button</a></p></p>
<strong>This is another strong tag</strong>
<hr />
<p><cite>This is cite</cite></p>
<blockquote cite="">This is a blockquote</blockquote>
<p><code>This is code</code></p>
<ul>
	<li>This is an li</li>
</ul>
<dl>
	<dd>This is a dd</dd>
	<dt>This is a dt</dt>
</dl>
<p><em>This is an em.</em></p>
```

_Note: Some of these elements are within `<p>` tags, which the WYSIWYG editor will add by default upon saving, feel free to remove them from the `<p>` containers and test them individually as well. You can do this by saving your page while still viewing in SOURCE view, it saves before the CKEDITOR has a chance to tamper with your code. If you have any questions on this please let me know._

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

### CODE REVIEW

File: docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss [hunk 1]
```diff
@@ -404,21 +404,70 @@
...
+  .hs-external-link,
+  .hs-file-link,
+  .hs-pdf-link,
+  .hs-private-link {
+    @include hb-text-reset;
+    font-style: italic;
     @include clearfix;
     line-height: 120%;
-    margin: hb-calculate-rems(16px) hb-calculate-rems(8px) hb-calculate-rems(8px);
+    margin-right: hb-calculate-rems(1.5px);
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss]: Let's add a comment here explaining what this is for.

File: docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss [hunk 2]
```diff
@@ -404,21 +404,70 @@
...
-    @include grid-media-min('sm') {
-      margin: hb-calculate-rems(24px) hb-calculate-rems(16px) hb-calculate-rems(12px);
-      width: 50%;
+    &::before,
+    &::after {
+      display: none !important;
     }
 
-    @include grid-media-min('md') {
-      width: 30%;
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss:421]: We need to add the width back in. I believe the card you are working on right now addresses refinements to the width. See lines 413 and 417 for other width breakpoints.

I know with the changes made to accommodate multiple paragraph elements, the widths declared above no longer work. Additional refactoring may be needed. Let's ask H&S if the width at 100% is ok for the work in this card and perhaps you can address the refactor in your current card.

**Review by heatherdesigns — COMMENTED** [review]
This is looking great! It looks like we will need to do some work around re-introducing the width. I left a comment in the code.

I also noticed that we need to add some styles in case some one applies the Splash font. It looks like it is serif while everything else is sans-serif. This is on Traditional. I didn't check Colorful.

splash

File: docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss (hunk 2 above)
Comment (rkelbel) [reply docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss:421]: This approach makes sense, removing the width on the <p> was essential to make this function inline, but we may be able to work out a better way of controlling the width on STN-660 with the smart trimming active as well.

I have sent a message in our external slack to clarify this approach with H&S.

File: docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss [hunk 3]
```diff
@@ -404,21 +404,70 @@
...
+  .hs-external-link,
+  .hs-file-link,
+  .hs-pdf-link,
+  .hs-private-link {
+    @include hb-text-reset;
+    font-style: italic;
     @include clearfix;
     line-height: 120%;
-    margin: hb-calculate-rems(16px) hb-calculate-rems(8px) hb-calculate-rems(8px);
+    margin-right: hb-calculate-rems(1.5px); // Adds spacing to elements for readability with spacing between inline text elements.
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss]: Comment added as requested.

File: docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss [hunk 4]
```diff
@@ -1,6 +1,61 @@
...
+    .hs-pdf-link,
+    .hs-private-link {
+      display: inline-block;
+    }
+
+    code,
+    h1,
+    h2,
+    h3,
+    .hs-font-splash {
```
Comment (rkelbel) [inline docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss:44]: Added Splash Font to the font-styles override, this particular class should be getting stripped out by the minimal WYSIWYG, but if it doesn't, it is covered.

Comment (heatherdesigns) [reply docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss:44]: Thank you!! 🏆

**Review by heatherdesigns — COMMENTED** [review]
I found a couple things that I missed the first time. Sorry!

 1. We need a hover state for when an anchor tag is located inside of a heading element. Right now the color of the text turns black on hover.

hover state of links in a heading element

2. The blockquote element is displaying the `"` symbol. It's located above the `o` in `caption` and is easier to see when you increase the screen size.

blockquote

It looks like you will need to update the styles on the paragraph `&:before` and `&:after` that sit inside of the block quote to not display the content.

before

3. Ideally we want everything to wrap. Right now each paragraph is still breaking into separate blocks.

dining couple pic

The last paragraph has plenty of room to wrap beside `SGS`.
summer film festival

I played around a bit to see if I could get the text to wrap. Here's a screenshot of the changes I made in my editor if you would like to test it.

Screen Shot 2021-01-25 at 9 05 51 PM

Screen Shot 2021-01-25 at 9 08 03 PM

Screen Shot 2021-01-25 at 9 08 15 PM

Let me know if it would be easier for me to commit the changes I made locally and push it up for you to check out. 👍 Hopefully those changes don't conflict in any way to the way you have in your other card. 🤞

File: docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss [hunk 5]
```diff
@@ -404,21 +404,70 @@
...
     text-align: right;
     float: right;
-    width: 90%;
+    color: white;
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss]: Let's use a color variable here.

Comment (rkelbel) [reply docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.text.scss]: Good catch.

File: docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss [hunk 6]
```diff
@@ -1,6 +1,65 @@
...
+    }
+
+    blockquote {
+      font-family: $hb-colorful-font--sans;
+
+      &:nth-child(n)::before,
+      &:nth-child(n)>p::before,
+      &:nth-child(n)>div::before,
+      &:nth-child(n)>p::after,
+      &:nth-child(n)>div::after {
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss]: Let's add some spaces to make our tests happy. :smile:

tests

Comment (heatherdesigns) [reply docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss]: I think we may want to add `&:nth-child(n)::after` as well.

You can also replace the `p` and `div` declarations with an `*` that would cover both instances.
So you could write, `&:nth-child(n) > *::before` and `&:nth-child(n) > *::after`. 
That one's your call. I'm cool with it it way. :tada:

**Comment by rkelbel** [comment]
> I found a couple things that I missed the first time. Sorry!
> 
> Let me know if it would be easier for me to commit the changes I made locally and push it up for you to check out. 👍 Hopefully those changes don't conflict in any way to the way you have in your other card. 🤞

Great changes Heather and good catch on the blockquote "'s, those were coming from the paragraph within the blockquote, which I initially missed, awesome catch and great feedback!

File: docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss (hunk 6 above)
Comment (rkelbel) [reply docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss]: Yep, caught that in my new push, should already be fixed up with a new fetch. 😄

File: docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss [hunk 7]
```diff
@@ -1,6 +1,64 @@
...
+    .hs-button,
+    .hs-button--big,
+    .hs-secondary-button,
+    .hs-more-link,
+    .hs-external-link,
+    .hs-file-link,
+    .hs-pdf-link,
+    .hs-private-link {
+      display: inline;
+      margin-right: hb-calculate-rems(1.5px); // Adds spacing to elements for readability with spacing between inline text elements.
```
Comment (heatherdesigns) [inline docroot/themes/humsci/humsci_basic/src/scss/utilities/_caption-credit.scss:37]: Excellent comment!!

**Review by heatherdesigns — APPROVED** [review]

**Pull request merged**